### ISSUE TITLE
TypeFactory use of `checkNull` shows up on profiles

### ISSUE THREAD

**Issue by jurgenvinju** [usethesource/vallang#233]
```
/**
	 * Construct a function type with labeled parameter and keyword field types
	 */
	public Type functionType(Type returnType, Type[] argumentTypes, String[] argumentLabels, Type[] keywordParameterTypes, String[] keywordParameterLabels) {
		checkNull((Object[]) argumentTypes);
		checkNull((Object[]) argumentLabels);
		checkNull((Object[]) keywordParameterTypes);
    	checkNull((Object[]) keywordParameterLabels);
		return getFromCache(new FunctionType(returnType, 
			(TupleType) getFromCache(new TupleTypeWithFieldNames(argumentTypes, argumentLabels)), 
			(TupleType) getFromCache(new TupleTypeWithFieldNames(keywordParameterTypes, keywordParameterLabels))
		));
	}
```

Given our use of the checker framework, and the possibility to put this in asserts, I wonder if we can either:
* remove all these checkNulls?
* or, make sure they are only activated in `-ea` mode, using `assert`? 

These `checkNull`'s showed up in a profile for testing the speed of transitive closure. Due to another optimization they have since dropped off, but still it seems like this is unnecessary overhead.

**Comment by DavyLandman** [comment]
1. do you have a case I can run that shows these in the profile? so I can reproduce it before I fix it?
2. I'm pretty sure these are not needed for CF, as the arguments are not marked `@Nullable`. Are these checks here since a user can pass null from the consumers side (if they aren't using CF?)? So it's proper argument validation? (and should we change the interface to mark those parameters as `@Nullable`?

Looking a bit more, it's our own function, that goes into the array, and loops over all of the entries and checks they are not null. If we don't want that guarantee: fine, we drop it, but this it not a CF concern, as it tends to be unsound for array manipulation. And again, our type doesn't say the array type can be nullable.

(These checks were introduced in this refactoring: https://github.com/usethesource/vallang/commit/de90bfa7e816ca0665516cfc3e6ee410fb320ea7 not as part of the move to CF.)

**Comment by jurgenvinju** [comment]
Indeed, these are not needed for CF. They were added for defensive programming before we had CF. The question is if we can replace them by CF guarantees or by asserts.

**Comment by jurgenvinju** [comment]
TypeFactory has these checks for all methods that construct types that have arguments.  They are thus always executed with every type that is constructed.

**Comment by DavyLandman** [comment]
There are some limitations with CF & arrays, see for example this entry in the documentation: https://checkerframework.org/manual/#nullness-arrays

**Comment by jurgenvinju** [comment]
Ok. In that case, let's rewrite this code to `assert checkNull(parameter)` then? This will make sure we check for nullness while running test code, and not while running production code.

**Comment by DavyLandman** [comment]
sounds like a plan to me :)

**Comment by jurgenvinju** [comment]
First #234 . That's an "interesting" one.

**Comment by jurgenvinju** [comment]
Having fun ;-)
